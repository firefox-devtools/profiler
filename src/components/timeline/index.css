/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.timelineOverflowEdgeIndicatorScrollbox {
  position: absolute;

  /* if the platform scrollbar ends up being larger than --vertical-scrollbar-reserved-width (which is unexpected), make sure there is no horizontal scrollbar */
  overflow: hidden auto;
  width: 100%;
  height: 100%;
  padding-left: var(--thread-label-column-width);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);
  background: linear-gradient(
    to left,
    white,
    #eee calc(var(--vertical-scrollbar-reserved-width) - 1px),
    var(--grey-30) 0,
    var(--grey-30) var(--vertical-scrollbar-reserved-width),
    transparent 0
  );
}

.timelineThreadList {
  width: calc(100vw - var(--vertical-scrollbar-reserved-width));
  padding: 0;
  margin: 0 0 0 calc(var(--thread-label-column-width) * -1);
  background-color: var(--grey-20);
  box-shadow: inset 0 1px var(--grey-30);
  list-style: none;
}

.timelineSettingsHiddenTracks {
  /* Make sure and override all rules that have to do with button defaults. */
  position: relative;
  width: var(--thread-label-column-width);

  /* 8px is intentional to align this text with the text above. */
  padding: 0 20px 0 8px;
  border: 0;

  /* Separator line between tracks menu button and timeline ruler */
  border-right: 1px solid var(--grey-30);
  background: none;
  color: inherit;
  font: inherit;
  text-align: left;
}

.timelineSettingsHiddenTracks:hover {
  background-color: rgb(0 0 0 / 0.1);
}

.timelineSettingsHiddenTracks:hover:active {
  background-color: rgb(0 0 0 / 0.2);
}

/* This is the dropdown arrow on the left of the button. */
.timelineSettingsHiddenTracks::before {
  position: absolute;
  top: 2px;
  right: 2px;
  border-top: 6px solid;
  border-right: 4px solid transparent;
  border-bottom: 0 solid transparent;
  border-left: 4px solid transparent;
  margin-top: 5px;
  margin-right: 5px;
  margin-left: 5px;
  color: var(--grey-60);
  content: '';

  /* Opt-out of forced colors so arrow is visible in High Contrast Mode */
  forced-color-adjust: none;
}

.timelineSettingsHiddenTracksNumber {
  font-weight: bold;
}

.timelineHeader {
  /* This should be on top of the overflow edge indicator that wraps the thread
   list, so border bottom will always be visible. */
  z-index: 1;
  display: flex;
  height: 20px;
  flex-direction: row;

  /* Timeline header content should end before the scrollbar but also border
   * bottom should extend to the whole timeline header. Because of that, we are
   * creating a padding and a minus margin with the scrollbar width to make
   * this work. */
  padding-right: var(--vertical-scrollbar-reserved-width);
  border-bottom: 1px solid var(--grey-30);
  margin-right: calc(var(--vertical-scrollbar-reserved-width) * -1);
  margin-left: calc(var(--thread-label-column-width) * -1);

  /* This div needs to be apaque so that it can properly hide the hidden tracks
   * container border below */
  background: white;
}

.tracksContainer {
  /* This should be below the timeline header. */
  z-index: 0;

  /* Moving the timeline 1 pixel below the timeline header. This way, the first
   * visible track's top border will be hidden as expected. */
  margin-top: -1px;
  margin-bottom: -1px;
}

@media (forced-colors: active) {
  .timelineSettingsHiddenTracks {
    border: 1px solid currentcolor;
    background-color: ButtonFace;
    color: ButtonText;
  }

  .timelineSettingsHiddenTracks:hover {
    background-color: SelectedItemText;
    color: SelectedItem;
  }

  .timelineSettingsHiddenTracks:hover:active {
    border-color: ButtonText;
  }

  .timelineSettingsHiddenTracks::before {
    color: currentcolor;
  }
}
