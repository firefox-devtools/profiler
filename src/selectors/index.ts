/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';
export * from './code';

import * as app from './app';
import {
  selectedThreadSelectors as selectedThread,
  selectedNodeSelectors as selectedNode,
} from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';
import * as code from './code';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  selectedNode,
  l10n,
  cpu,
  code,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
