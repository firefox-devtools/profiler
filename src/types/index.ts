/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// TypeScript types
export * from './actions';
export * from './gecko-profile';
export * from './indexeddb';
export * from './markers';
export * from './profile-derived';
export * from './profile';
export * from './state';
export * from './store';
export * from './symbolication';
export * from './transforms';
export * from './units';
export * from './utils';
