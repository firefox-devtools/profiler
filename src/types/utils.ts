/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
