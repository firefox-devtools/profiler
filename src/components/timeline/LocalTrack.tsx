/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { PureComponent } from 'react';
import { Localized } from '@fluent/react';
import classNames from 'classnames';
import {
  changeRightClickedTrack,
  selectTrackWithModifiers,
  hideLocalTrack,
} from 'firefox-profiler/actions/profile-view';
import { assertExhaustiveCheck } from 'firefox-profiler/utils/types';
import { ContextMenuTrigger } from 'firefox-profiler/components/shared/ContextMenuTrigger';
import {
  getSelectedThreadIndexes,
  getSelectedTab,
  getHiddenLocalTracks,
} from 'firefox-profiler/selectors/url-state';
import explicitConnect from 'firefox-profiler/utils/connect';
import {
  getLocalTrackName,
  getCounterSelectors,
} from 'firefox-profiler/selectors/profile';
import { getThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import { TimelineTrackThread } from './TrackThread';
import { TrackEventDelay } from './TrackEventDelay';
import { TrackNetwork } from './TrackNetwork';
import { TrackMemory } from './TrackMemory';
import { TrackBandwidth } from './TrackBandwidth';
import { TrackIPC } from './TrackIPC';
import { TrackProcessCPU } from './TrackProcessCPU';
import { TrackPower } from './TrackPower';
import { getTrackSelectionModifiers } from 'firefox-profiler/utils';
import type {
  TrackReference,
  Pid,
  TrackIndex,
  LocalTrack,
} from 'firefox-profiler/types';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';
import { TrackCustomMarker } from './TrackCustomMarker';

type OwnProps = {
  readonly pid: Pid;
  readonly localTrack: LocalTrack;
  readonly trackIndex: TrackIndex;
  readonly style?: React.CSSProperties /* This is used by Reorderable */;
  readonly setIsInitialSelectedPane: (value: boolean) => void;
};

type StateProps = {
  readonly trackName: string;
  readonly isSelected: boolean;
  readonly isHidden: boolean;
  readonly titleText: string | undefined;
};

type DispatchProps = {
  readonly changeRightClickedTrack: typeof changeRightClickedTrack;
  readonly selectTrackWithModifiers: typeof selectTrackWithModifiers;
  readonly hideLocalTrack: typeof hideLocalTrack;
};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

class LocalTrackComponent extends PureComponent<Props> {
  _onContextMenu = () => {
    // Notify the redux store that this was right clicked.
    this.props.changeRightClickedTrack(this._getTrackReference());
  };

  _selectCurrentTrack = (
    event: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
  ) => {
    this.props.selectTrackWithModifiers(
      this._getTrackReference(),
      getTrackSelectionModifiers(event)
    );
  };

  _getTrackReference(): TrackReference {
    const { pid, trackIndex } = this.props;
    return { type: 'local', pid, trackIndex };
  }

  _hideCurrentTrack = (
    event: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
  ) => {
    const { pid, trackIndex, hideLocalTrack } = this.props;
    hideLocalTrack(pid, trackIndex);
    event.stopPropagation();
  };

  renderTrack() {
    const { localTrack, trackName } = this.props;
    switch (localTrack.type) {
      case 'thread':
        return (
          <TimelineTrackThread
            threadsKey={localTrack.threadIndex}
            trackType="expanded"
            trackName={trackName}
          />
        );
      case 'network':
        return <TrackNetwork threadIndex={localTrack.threadIndex} />;
      case 'memory':
        return <TrackMemory counterIndex={localTrack.counterIndex} />;
      case 'bandwidth':
        return <TrackBandwidth counterIndex={localTrack.counterIndex} />;
      case 'ipc':
        return <TrackIPC threadIndex={localTrack.threadIndex} />;
      case 'event-delay':
        return <TrackEventDelay threadIndex={localTrack.threadIndex} />;
      case 'process-cpu':
        return <TrackProcessCPU counterIndex={localTrack.counterIndex} />;
      case 'power':
        return <TrackPower counterIndex={localTrack.counterIndex} />;
      case 'marker':
        return (
          <TrackCustomMarker
            threadIndex={localTrack.threadIndex}
            markerSchema={localTrack.markerSchema}
            markerName={localTrack.markerName}
          />
        );
      default:
        console.error('Unhandled localTrack type', localTrack as never);
        return null;
    }
  }

  override componentDidMount() {
    const { isSelected } = this.props;
    if (isSelected) {
      this.props.setIsInitialSelectedPane(true);
    }
  }

  override render() {
    const { isSelected, isHidden, titleText, trackName, style } = this.props;

    if (isHidden) {
      // If this global track is hidden, render out a stub element so that the
      // Reorderable Component still works across all the tracks.
      return <li className="timelineTrackHidden" />;
    }

    return (
      <li className="timelineTrack timelineTrackLocal" style={style}>
        {/* This next div is used to mirror the structure of the TimelineGlobalTrack */}
        <ContextMenuTrigger
          id="TimelineTrackContextMenu"
          renderTag="div"
          attributes={{
            className: classNames('timelineTrackRow timelineTrackLocalRow', {
              selected: isSelected,
            }),
            onContextMenu: this._onContextMenu,
          }}
        >
          <div
            className="timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
            title={titleText}
            onClick={this._selectCurrentTrack}
          >
            <button type="button" className="timelineTrackNameButton">
              {trackName}
            </button>
            <Localized id="TrackNameButton--hide-track" attrs={{ title: true }}>
              <button
                type="button"
                className="timelineTrackCloseButton"
                title="Hide track"
                onClick={this._hideCurrentTrack}
              />
            </Localized>
          </div>
          <div
            className="timelineTrackTrack"
            onClick={this._selectCurrentTrack}
          >
            {this.renderTrack()}
          </div>
        </ContextMenuTrigger>
      </li>
    );
  }
}

export const TimelineLocalTrack = explicitConnect<
  OwnProps,
  StateProps,
  DispatchProps
>({
  mapStateToProps: (state, { pid, localTrack, trackIndex }) => {
    // These get assigned based on the track type.
    let titleText;
    let isSelected = false;

    const selectedThreadIndexes = getSelectedThreadIndexes(state);

    // Run different selectors based on the track type.
    switch (localTrack.type) {
      case 'thread': {
        // Look up the thread information for the process if it exists.
        const threadIndex = localTrack.threadIndex;
        const selectedTab = getSelectedTab(state);
        const selectors = getThreadSelectors(threadIndex);
        isSelected =
          selectedThreadIndexes.has(threadIndex) &&
          selectedTab !== 'network-chart';
        titleText = selectors.getThreadProcessDetails(state);
        break;
      }
      case 'network': {
        const threadIndex = localTrack.threadIndex;
        const selectedTab = getSelectedTab(state);
        isSelected =
          selectedThreadIndexes.has(threadIndex) &&
          selectedTab === 'network-chart';
        break;
      }
      case 'marker':
      case 'ipc': {
        const threadIndex = localTrack.threadIndex;
        const selectedTab = getSelectedTab(state);
        isSelected =
          selectedThreadIndexes.has(threadIndex) &&
          selectedTab === 'marker-chart';
        break;
      }
      case 'event-delay': {
        // Look up the thread information for the process if it exists.
        const threadIndex = localTrack.threadIndex;
        const selectors = getThreadSelectors(threadIndex);
        isSelected = selectedThreadIndexes.has(threadIndex);
        titleText =
          'Event Delay of ' + selectors.getThreadProcessDetails(state);
        break;
      }
      case 'memory':
      case 'bandwidth':
      case 'process-cpu':
      case 'power': {
        titleText = getCounterSelectors(localTrack.counterIndex).getDescription(
          state
        );
        break;
      }
      default:
        throw assertExhaustiveCheck(localTrack, `Unhandled LocalTrack type.`);
    }

    return {
      trackName: getLocalTrackName(state, pid, trackIndex),
      titleText,
      isSelected,
      isHidden: getHiddenLocalTracks(state, pid).has(trackIndex),
    };
  },
  mapDispatchToProps: {
    changeRightClickedTrack,
    selectTrackWithModifiers,
    hideLocalTrack,
  },
  component: LocalTrackComponent,
});
