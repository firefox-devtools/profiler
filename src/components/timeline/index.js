/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import { FullTimeline } from 'firefox-profiler/components/timeline/FullTimeline';

type TimelineProps = {||};

export class Timeline extends PureComponent<TimelineProps> {
  // This may contain a function that's called whenever we want to remove the
  // "wheel" listener.
  _removeWheelListener: null | (() => mixed) = null;

  // This effectively disable the pinch-to-zoom as well as ctrl+mousewheel
  // gestures. Indeed in the timeline it is confusing. In the future we'll want
  // to couple this with the preview selection like in the Viewport HOC.
  preventPinchToZoom(e: WheelEvent) {
    if (e.ctrlKey) {
      e.preventDefault();
    }
  }

  // This will be registered as a ref property to the DOM element displaying the
  // tracks. We use this solution of registering the wheel event in this ref
  // listener because:
  // * we can't use React's event handling, because it doesn't allow us to use
  //   the "passive: false" way of registering the event handler. But we need this
  //   if we want to be able to prevent the default action of page zooming.
  // * we want to be sure to register it whenever the element changes.
  _onTimelineMountWithRef = (ref: HTMLElement | null) => {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }

    if (!ref) {
      return;
    }

    // without pinning to a const variable, Flow isn't sure that we don't change
    // the `ref` variable in some of the function calls below, and therefore
    // that it won't be null.
    const existingRef = ref;

    // Disable pinch-to-zoom and ctrl + wheel otherwise, on the timeline.
    // Indeed the users are used to this gesture  to zoom in in our charts, and
    // may use the same gesture elsewhere because of their habits, however this
    // doesn't do what they expect and instead zooms in the page, which is distracting.
    existingRef.addEventListener('wheel', this.preventPinchToZoom, {
      passive: false,
    });

    this._removeWheelListener = () => {
      existingRef.removeEventListener('wheel', this.preventPinchToZoom, {
        passive: false,
      });
    };
  };

  componentWillUnmount() {
    if (this._removeWheelListener) {
      this._removeWheelListener();
      this._removeWheelListener = null;
    }
  }

  render() {
    return <FullTimeline innerElementRef={this._onTimelineMountWithRef} />;
  }
}
