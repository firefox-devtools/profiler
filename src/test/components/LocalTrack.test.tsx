/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import type {
  TrackReference,
  Store,
  ThreadIndex,
  LocalTrack,
} from 'firefox-profiler/types';
import { Provider } from 'react-redux';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import {
  changeSelectedThreads,
  hideLocalTrack,
  showLocalTrack,
} from '../../actions/profile-view';
import { TimelineLocalTrack } from '../../components/timeline/LocalTrack';
import {
  getRightClickedTrack,
  getLocalTrackFromReference,
  getProfile,
} from '../../selectors/profile';
import { ensureExists } from '../../utils/types';
import { getFirstSelectedThreadIndex } from '../../selectors/url-state';
import { autoMockCanvasContext } from '../fixtures/mocks/canvas-context';
import {
  getNetworkTrackProfile,
  getIPCTrackProfile,
} from '../fixtures/profiles/processed-profile';
import {
  getProfileWithNiceTracks,
  getStoreWithMemoryTrack,
} from '../fixtures/profiles/tracks';
import { storeWithProfile } from '../fixtures/stores';
import { fireFullClick, fireFullContextMenu } from '../fixtures/utils';
import { autoMockElementSize } from '../fixtures/mocks/element-size';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { autoMockIntersectionObserver } from '../fixtures/mocks/intersection-observer';

// In getProfileWithNiceTracks, the two pids are 111 and 222 for the
// "GeckoMain process" and "GeckoMain tab" respectively. Use 222 since it has
// local tracks.
const PID = '222';

describe('timeline/LocalTrack', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: 400, height: 400 });
  autoMockIntersectionObserver();

  describe('with a thread track', function () {
    it('matches the snapshot of a local track', () => {
      const { container } = setupThreadTrack();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('has the correct selectors into useful parts of the component', function () {
      const { getLocalTrackLabel, getLocalTrackRow } = setupThreadTrack();
      expect(getLocalTrackLabel()).toHaveTextContent('DOM Worker');
      expect(getLocalTrackRow()).toBeTruthy();
    });

    it('starts out not being selected', function () {
      const { getState, threadIndex, trackReference, getLocalTrackRow } =
        setupThreadTrack();
      expect(getRightClickedTrack(getState())).not.toEqual(trackReference);
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
      expect(getLocalTrackRow()).not.toHaveClass('selected');
    });

    it('can select a thread by clicking the label', () => {
      const { getState, getLocalTrackLabel, threadIndex, getLocalTrackRow } =
        setupThreadTrack();
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
      fireFullClick(getLocalTrackLabel());
      expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
      expect(getLocalTrackRow()).toHaveClass('selected');
    });

    it('can right click a thread on the label', () => {
      const { getState, getLocalTrackLabel, threadIndex, trackReference } =
        setupThreadTrack();

      fireFullContextMenu(getLocalTrackLabel());
      expect(getRightClickedTrack(getState())).toEqual(trackReference);
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
    });

    it('can select a thread by clicking the track content', () => {
      const { getState, getLocalTrackContent, threadIndex } =
        setupThreadTrack();
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
      fireFullClick(getLocalTrackContent());
      expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
    });

    it('can right click a thread on the track content', () => {
      const { getState, getLocalTrackContent, threadIndex, trackReference } =
        setupThreadTrack();

      fireFullContextMenu(getLocalTrackContent());
      expect(getRightClickedTrack(getState())).toEqual(trackReference);
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
    });

    it('will render a stub div if the track is hidden', () => {
      const { container, pid, trackReference, dispatch } = setupThreadTrack();
      act(() => {
        dispatch(hideLocalTrack(pid, trackReference.trackIndex));
      });
      expect(container.querySelector('.timelineTrackHidden')).toBeTruthy();
      expect(container.querySelector('.timelineTrack')).toBeFalsy();
    });
  });

  describe('with a network track', function () {
    it('has correctly renders the network label', function () {
      const { getLocalTrackLabel } = setupWithNetworkProfile();
      expect(getLocalTrackLabel()).toHaveTextContent('Network');
    });

    it('matches the snapshot of the network track', () => {
      const { container } = setupWithNetworkProfile();
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('with a memory track', function () {
    it('correctly renders the network label', function () {
      const { getLocalTrackLabel } = setupWithMemory();
      expect(getLocalTrackLabel()).toHaveTextContent('Memory');
    });

    it('matches the snapshot of the memory track', () => {
      const { container } = setupWithMemory();
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('with an IPC track', function () {
    it('appears hidden by default', function () {
      const { container } = setupWithIPC();
      expect(container.querySelector('.timelineTrackHidden')).toBeTruthy();
      expect(container.querySelector('.timelineTrack')).toBeFalsy();
    });

    it('can be shown', function () {
      const { dispatch, pid, trackReference, container } = setupWithIPC();

      // First check that the IPC track is hidden by default.
      expect(container.querySelector('.timelineTrackHidden')).toBeTruthy();
      expect(container.querySelector('.timelineTrack')).toBeFalsy();

      // Now make it visible and check it.
      act(() => {
        dispatch(showLocalTrack(pid, trackReference.trackIndex));
      });
      expect(container.querySelector('.timelineTrackHidden')).toBeFalsy();
      expect(container.querySelector('.timelineTrack')).toBeTruthy();
    });

    it('correctly renders the IPC label', function () {
      const { dispatch, pid, trackReference, getLocalTrackLabel } =
        setupWithIPC();
      act(() => {
        dispatch(showLocalTrack(pid, trackReference.trackIndex));
      });
      expect(getLocalTrackLabel()).toHaveTextContent('IPC â€” Empty');
    });

    it('matches the snapshot of the IPC track', () => {
      const { pid, dispatch, trackReference, container, flushRafCalls } =
        setupWithIPC();
      act(() => {
        dispatch(showLocalTrack(pid, trackReference.trackIndex));
      });
      flushRafCalls();
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

function setup(
  store: Store,
  trackReference: TrackReference,
  localTrack: LocalTrack,
  threadIndex: ThreadIndex
) {
  const { getState, dispatch } = store;
  const setIsInitialSelectedPane = () => {};
  // The assertions are simpler if this thread is not already selected.
  dispatch(changeSelectedThreads(new Set([threadIndex + 1])));

  // WithSize uses requestAnimationFrame
  const flushRafCalls = mockRaf();
  const renderResult = render(
    <Provider store={store}>
      <TimelineLocalTrack
        pid={PID}
        localTrack={localTrack}
        trackIndex={trackReference.trackIndex}
        setIsInitialSelectedPane={setIsInitialSelectedPane}
      />
    </Provider>
  );
  flushRafCalls();

  const { container } = renderResult;

  const getLocalTrackLabel = () =>
    ensureExists(
      container.querySelector('.timelineTrackLabel'),
      `Couldn't find the track label with selector .timelineTrackLabel`
    ) as HTMLElement;
  const getLocalTrackContent = () =>
    ensureExists(
      container.querySelector('.timelineTrackTrack'),
      `Couldn't find the track content with selector .timelineTrackTrack`
    ) as HTMLElement;
  const getLocalTrackRow = () =>
    ensureExists(
      container.querySelector('.timelineTrackLocalRow'),
      `Couldn't find the track local row with selector .timelineTrackLocalRow`
    ) as HTMLElement;

  return {
    ...renderResult,
    dispatch,
    getState,
    profile: getProfile(getState()),
    store,
    trackReference,
    threadIndex,
    pid: PID,
    getLocalTrackLabel,
    getLocalTrackContent,
    getLocalTrackRow,
    flushRafCalls,
  };
}

/**
 *  getProfileWithNiceTracks() looks like: [
 *    'show [thread GeckoMain default]',
 *    'show [thread GeckoMain tab]',
 *    '  - show [thread DOM Worker]',         <- use this local track.
 *    '  - show [thread Style]',
 *  ]
 */
function setupThreadTrack() {
  // Select the first local track index, which should be the one noted in the
  // the comment above.
  const trackIndex = 0;
  const profile = getProfileWithNiceTracks();
  const store = storeWithProfile(profile);
  const trackReference = { type: 'local' as const, pid: PID, trackIndex };
  const localTrack = getLocalTrackFromReference(
    store.getState(),
    trackReference
  );
  if (localTrack.type !== 'thread') {
    throw new Error('Expected a thread track.');
  }
  return setup(store, trackReference, localTrack, localTrack.threadIndex);
}

function setupWithNetworkProfile() {
  // Select the 2nd track, which will be the network track, while the first is the
  // main thread.
  const trackIndex = 1;
  const profile = getNetworkTrackProfile();
  profile.threads[0].pid = PID;

  const store = storeWithProfile(profile);
  const trackReference = { type: 'local' as const, pid: PID, trackIndex };
  const localTrack = getLocalTrackFromReference(
    store.getState(),
    trackReference
  );
  if (localTrack.type !== 'network') {
    throw new Error('Expected a network track.');
  }
  return setup(store, trackReference, localTrack, localTrack.threadIndex);
}

/**
 * Set up a profile with a memory counter.
 */
function setupWithMemory() {
  const { store, trackReference, localTrack, threadIndex } =
    getStoreWithMemoryTrack(PID);
  return setup(store, trackReference, localTrack, threadIndex);
}

/**
 * Set up a profile with an IPC message track.
 */
function setupWithIPC() {
  // Select the 2nd track, which will be the IPC track.
  const trackIndex = 1;
  const profile = getIPCTrackProfile();
  profile.threads[0].pid = PID;

  const store = storeWithProfile(profile);
  const trackReference = { type: 'local' as const, pid: PID, trackIndex };
  const localTrack = getLocalTrackFromReference(
    store.getState(),
    trackReference
  );
  if (localTrack.type !== 'ipc') {
    throw new Error('Expected an IPC track.');
  }
  return setup(store, trackReference, localTrack, localTrack.threadIndex);
}
