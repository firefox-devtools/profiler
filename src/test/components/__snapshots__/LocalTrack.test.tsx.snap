// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`timeline/LocalTrack with a memory track matches the snapshot of the memory track 1`] = `
<li
  class="timelineTrack timelineTrackLocal"
>
  <div
    class="timelineTrackRow timelineTrackLocalRow"
  >
    <div
      class="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
      title="My Description"
    >
      <button
        class="timelineTrackNameButton"
        type="button"
      >
        Memory
      </button>
      <button
        class="timelineTrackCloseButton"
        title="Hide track"
        type="button"
      />
    </div>
    <div
      class="timelineTrackTrack"
    >
      <div
        class="timelineTrackMemory"
        style="height: 40px; --graph-height: 25px; --markers-height: 15px;"
      >
        <div
          class="timelineMarkers timelineMarkersMemory"
          data-testid="TimelineMarkersMemory"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <div>
              <canvas
                class="timelineMarkersCanvas"
                height="400"
                width="400"
              />
            </div>
          </div>
        </div>
        <div
          class="timelineTrackMemoryGraph"
        >
          <div>
            <canvas
              class="timelineTrackMemoryCanvas"
              height="25"
              width="400"
            />
          </div>
          <div
            class="timelineEmptyThreadIndicator"
          />
        </div>
      </div>
    </div>
  </div>
</li>
`;

exports[`timeline/LocalTrack with a network track matches the snapshot of the network track 1`] = `
<li
  class="timelineTrack timelineTrackLocal"
>
  <div
    class="timelineTrackRow timelineTrackLocalRow"
  >
    <div
      class="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
    >
      <button
        class="timelineTrackNameButton"
        type="button"
      >
        Network
      </button>
      <button
        class="timelineTrackCloseButton"
        title="Hide track"
        type="button"
      />
    </div>
    <div
      class="timelineTrackTrack"
    >
      <div
        class="timelineTrackNetwork"
        style="height: 35px;"
      >
        <div
          class="react-contextmenu-wrapper treeViewContextMenu"
        >
          <canvas
            class="timelineTrackNetworkCanvas"
            height="35"
            width="400"
          />
          <div
            class="timelineVerticalIndicatorsLine"
            data-marker-index="10"
            data-testid="vertical-indicator-line"
            style="--vertical-indicator-color: var(--red-60); left: 200px;"
          />
          <div
            class="timelineVerticalIndicatorsLine"
            data-marker-index="11"
            data-testid="vertical-indicator-line"
            style="--vertical-indicator-color: #000; left: 200px;"
          />
          <div
            class="timelineVerticalIndicatorsLine"
            data-marker-index="12"
            data-testid="vertical-indicator-line"
            style="--vertical-indicator-color: var(--blue-50); left: 266.6666666666667px;"
          />
          <div
            class="timelineVerticalIndicatorsLine"
            data-marker-index="13"
            data-testid="vertical-indicator-line"
            style="--vertical-indicator-color: var(--grey-40); left: 266.6666666666667px;"
          />
          <div
            class="timelineVerticalIndicatorsLine"
            data-marker-index="14"
            data-testid="vertical-indicator-line"
            style="--vertical-indicator-color: var(--magenta-60); left: 333.33333333333337px;"
          />
        </div>
      </div>
    </div>
  </div>
</li>
`;

exports[`timeline/LocalTrack with a thread track matches the snapshot of a local track 1`] = `
<li
  class="timelineTrack timelineTrackLocal"
>
  <div
    class="timelineTrackRow timelineTrackLocalRow"
  >
    <div
      class="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
      title="DOM Worker
Thread: \\"DOM Worker\\" (33)
Process: \\"tab\\" (222)"
    >
      <button
        class="timelineTrackNameButton"
        type="button"
      >
        DOM Worker
      </button>
      <button
        class="timelineTrackCloseButton"
        title="Hide track"
        type="button"
      />
    </div>
    <div
      class="timelineTrackTrack"
    >
      <div
        class="timelineTrackThread expanded"
      >
        <div
          class="timelineMarkers"
          data-testid="TimelineMarkersJank"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <div>
              <canvas
                class="timelineMarkersCanvas"
                height="400"
                width="400"
              />
            </div>
          </div>
        </div>
        <div
          class="threadActivityGraph"
        >
          <div>
            <canvas
              class="threadActivityGraphCanvas threadActivityGraphCanvas"
              height="400"
              width="400"
            >
              <h2>
                Activity Graph for 
                DOM Worker
              </h2>
              <p>
                This graph shows a visual chart of thread activity.
              </p>
            </canvas>
          </div>
        </div>
        <div
          class="threadSampleGraph"
        >
          <div>
            <canvas
              class="threadSampleGraphCanvas threadSampleGraphCanvas"
              height="400"
              width="400"
            >
              <h2>
                Stack Graph for 
                DOM Worker
              </h2>
              <p>
                This graph charts the stack height of each sample.
              </p>
            </canvas>
          </div>
        </div>
        <div
          class="timelineEmptyThreadIndicator"
        />
      </div>
    </div>
  </div>
</li>
`;

exports[`timeline/LocalTrack with an IPC track matches the snapshot of the IPC track 1`] = `
<li
  class="timelineTrack timelineTrackLocal"
>
  <div
    class="timelineTrackRow timelineTrackLocalRow"
  >
    <div
      class="react-contextmenu-wrapper timelineTrackLabel timelineTrackLocalLabel timelineTrackLocalGrippy"
    >
      <button
        class="timelineTrackNameButton"
        type="button"
      >
        IPC â€” Empty
      </button>
      <button
        class="timelineTrackCloseButton"
        title="Hide track"
        type="button"
      />
    </div>
    <div
      class="timelineTrackTrack"
    >
      <div
        style="height: 25px; --markers-height: 25px;"
      >
        <div
          class="timelineMarkers timelineMarkersIPC"
          data-testid="TimelineMarkersIPC"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <div>
              <canvas
                class="timelineMarkersCanvas"
                height="400"
                width="400"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</li>
`;
